src/instrumentation/podmutator.go

```diff
@@ -35,150 +42,244 @@ var (
+// SecretReplicator is used to copy secrets from one namespace to another
+type SecretReplicator interface {
+	ReplicateSecret(ctx context.Context, ns corev1.Namespace, pod corev1.Pod, operatorNamespace string, secretName string) error
+}
 
-	// once it's been determined that instrumentation is desired, none exists yet, and we know which instance it should talk to,
-	// we should inject the instrumentation.
-	modifiedPod := pod
-	for _, currentContainer := range strings.Split(targetContainers, ",") {
-		modifiedPod = pm.sdkInjector.inject(ctx, insts, ns, modifiedPod, strings.TrimSpace(currentContainer))
-	}
+// NewrelicSecretReplicator is the base struct used for copying the secrets
+type NewrelicSecretReplicator struct {
+	client client.Client
+	logger logr.Logger
+}
 
-	return modifiedPod, nil
+// NewNewrelicSecretReplicator is the constructor for copying secrets
+func NewNewrelicSecretReplicator(logger logr.Logger, client client.Client) *NewrelicSecretReplicator {
+	return &NewrelicSecretReplicator{client: client, logger: logger}
 }
 
-func (pm *instPodMutator) getInstrumentationInstance(ctx context.Context, ns corev1.Namespace, pod corev1.Pod, instAnnotation string) (*v1alpha1.Instrumentation, error) {
-	instValue := annotationValue(ns.ObjectMeta, pod.ObjectMeta, instAnnotation)
+// ReplicateSecret is used to copy the secret from the operator namespace to the pod namespace if the secret doesn't already exist
+func (sr *NewrelicSecretReplicator) ReplicateSecret(ctx context.Context, ns corev1.Namespace, pod corev1.Pod, operatorNamespace string, secretName string) error {
+	logger := sr.logger.WithValues("namespace", pod.Namespace, "name", pod.Name)
 
-	if len(instValue) == 0 || strings.EqualFold(instValue, "false") {
-		return nil, nil
-	}
+	var secret corev1.Secret
 
-	if strings.EqualFold(instValue, "true") {
-		return pm.selectInstrumentationInstanceFromNamespace(ctx, ns)
+	if secretName == "" {
+		secretName = DefaultLicenseKeySecretName
 	}
 
-	var instNamespacedName types.NamespacedName
-	if instNamespace, instName, namespaced := strings.Cut(instValue, "/"); namespaced {
-		instNamespacedName = types.NamespacedName{Name: instName, Namespace: instNamespace}
-	} else {
-		instNamespacedName = types.NamespacedName{Name: instValue, Namespace: ns.Name}
+	err := sr.client.Get(ctx, client.ObjectKey{Namespace: ns.Name, Name: secretName}, &secret)
+	if err == nil {
+		logger.Info("secret already exists")
+		return nil
 	}
+	if err != nil && !apierrors.IsNotFound(err) {
+		logger.Error(err, "failed to check for existing secret in pod namespace")
+		return err
+	}
+	logger.Info("replicating secret to pod namespace")
 
-	nrInst := &v1alpha1.Instrumentation{}
-	err := pm.Client.Get(ctx, instNamespacedName, nrInst)
+	err = sr.client.Get(ctx, client.ObjectKey{Namespace: operatorNamespace, Name: secretName}, &secret)
 	if err != nil {
-		return nil, err
+		logger.Error(err, "failed to retrieve the secret from operator namespace")
+		return err
```

Comment:
nit : Let's do something like this in such scenarios.

```
if err := sr.client.Get(ctx, client.ObjectKey{Namespace: operatorNamespace, Name: secretName}, &secret); err != nil {
		logger.Error(err, "failed to retrieve the secret from operator namespace")
		return err
	}
```

Reply:
fixed